File: src/data_fetcher.jl
module DataFetcher

using Dates
using HTTP
using JSON
using DataFrames
using CSV

export fetch_data, save_data, load_data

"""
fetch historical price data from yahoo finance for a given ticker and date range.
    parameters:
    - ticker: stock symbol (default: "^GSPC" for S&P 500)
    - start_date: start date (default: 5 years ago)
    - end_date: end date (default: today)
    - interval: data interval (e.g, 1d, 1wk, 1mo, default: 1d)
returns a dataframe with historical price data.
"""
function fetch_data(;
    ticker::String = "^GSPC",
    start_date::Union{Date, DateTime} = Dates.today() - Dates.Year(5),
    end_date::Union{Date, DateTime} = Dates.today(),
    interval::String = "1d"
)
    # change Date to DateTime
    start_dt = start_date isa Date ? DateTime(start_date) : start_date
    end_dt = end_date isa Date ? DateTime(end_date) : end_date

    # change Datetime to unix time
    start_date_unix = Int(floor(datetime2unix(start_dt)))
    end_date_unix = Int(floor(datetime2unix(end_dt)))

    # create url to data
    base_url = "https://query1.finance.yahoo.com/v8/finance/chart/"
    params = "?period1=$(start_date_unix)&period2=$(end_date_unix)&interval=$(interval)&events=history"
    url = base_url * HTTP.escapeuri(ticker) * params

    try
        # make http request (with user agent to make requests more reliable)
        response = HTTP.get(url, headers=Dict("User-Agent" => "Mozilla/5.0"))

        # parse json response
        data = JSON.parse(String(response.body))

        # extract data
        result = data["chart"]["result"][1]
        timestamps = result["timestamp"]
        quote_data = result["indicators"]["quote"][1]

        # create data frame
        df = DataFrame(
            date = unix2datetime.(timestamps),
            open = get(quote_data, "open", fill(missing, length(timestamps))),
            high = get(quote_data, "high", fill(missing, length(timestamps))),
            low = get(quote_data, "low", fill(missing, length(timestamps))),
            close = get(quote_data, "close", fill(missing, length(timestamps))),
            volume = get(quote_data, "volume", fill(missing, length(timestamps)))
        )

        # handle any missing values
        df = dropmissing(df, :close)

        # sort by date with most recent data on top
        sort!(df, :date, rev = true)

        return df
    catch er
        error("Error fetching data: $(sprint(showerror, er))")
    end
end

"""
save data to a csv file.
"""
function save_data(
    df::DataFrame,
    ticker::String;
    path::String = "./data"
)
    # ensure directory exists
    mkpath(path)
    
    # create filename with ticker and date range
    filename = joinpath(path, "$(ticker)_$(Dates.format(minimum(df.date), "yyyymmdd"))_$(Dates.format(maximum(df.date), "yyyymmdd")).csv")

    try
        # write to file
        CSV.write(filename, df)
        
        return filename
    catch er
        error("Error saving data: $(sprint(showerror, er))")
    end
end

"""
load data from a csv file.
"""
function load_data(filename::String)
    try
        # read from file
        df = CSV.read(filename, DataFrame)
        
        # ensure date column is DateTime
        if :date in names(df) && eltype(df.date) <: AbstractString
            df.date = DateTime.(df.date)
        end

        return df
    catch er
        error("Error loading data: $(sprint(showerror, er))")
    end
end

end

File: src/data_fetcher.py
# data_fetcher.py
# Minimal data fetcher for Djanus investment strategy system
#
# Provides DataFetcher class for retrieving financial data from Yahoo Finance
# Handles single tickers with unified interface
# Returns raw Polars DataFrames for processing by DataProcessor
# Supports optional saving to CSV/Parquet files
# Can load existing files if they match parameters
#
# Main methods:
# - get_data(): retrieve raw market data
# - load_data(): load existing data from file
# - save_data(): persist DataFrame to file
# - fetch_data(): complete pipeline (check existing -> load or download -> optional save)

from typing import Union, Optional
from datetime import datetime, timedelta
from pathlib import Path
from enum import Enum
import polars as pl
import yfinance as yf
import logging

# Configure logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

# Type aliases
DateLike = Union[str, datetime]


class TimeInterval(Enum):
    """Supported time intervals for financial data."""

    DAILY = "1d"
    WEEKLY = "1wk"
    MONTHLY = "1mo"
    QUARTERLY = "3mo"


class FileFormat(Enum):
    """Supported file formats for data storage."""

    CSV = "csv"
    PARQUET = "parquet"


class DataFetcher:
    """Minimal data fetcher for Yahoo Finance data using yfinance."""

    # Helper methods
    def _parse_date(self, date: Optional[DateLike], days_back: int = 365) -> str:
        """
        Convert date to YYYY-MM-DD string format.

        Args:
            date: Date as string 'YYYY-MM-DD', datetime object, or None
            days_back: Days back from now for None dates (0 for current time)

        Returns:
            Date string in YYYY-MM-DD format
        """
        if date is None:
            base_date = (
                datetime.now()
                if days_back == 0
                else datetime.now() - timedelta(days=days_back)
            )
            return base_date.strftime("%Y-%m-%d")
        elif isinstance(date, str):
            try:
                datetime.strptime(date, "%Y-%m-%d")
                return date
            except ValueError:
                raise ValueError(f"Invalid date format '{date}'. Expected YYYY-MM-DD")
        else:
            return date.strftime("%Y-%m-%d")

    # Main methods
    def get_data(
        self,
        ticker: str = "^GSPC",
        start_date: Optional[DateLike] = None,
        end_date: Optional[DateLike] = None,
        interval: TimeInterval = TimeInterval.DAILY,
    ) -> pl.DataFrame:
        """
        Fetch historical data for a single ticker.

        Args:
            ticker: Ticker symbol (e.g., 'SPY', '^GSPC')
            start_date: Start date as string 'YYYY-MM-DD' or datetime object
            end_date: End date as string 'YYYY-MM-DD' or datetime object
            interval: Data interval from TimeInterval enum

        Returns:
            Polars DataFrame with raw OHLCV data and interval metadata
        """
        start_str = self._parse_date(start_date, 365)
        end_str = self._parse_date(end_date, 0)

        logger.info(f"Fetching {ticker} from {start_str} to {end_str}")

        try:
            ticker_obj = yf.Ticker(ticker)
            data = ticker_obj.history(
                start=start_str, end=end_str, interval=interval.value
            )
        except Exception as e:
            raise ValueError(f"Failed to fetch data for {ticker}: {e}")

        if data.empty:
            raise ValueError(f"No data found for {ticker}")

        # Convert to Polars and add metadata
        df = pl.from_pandas(data.reset_index())
        df = df.with_columns(
            [
                pl.col("Date").cast(pl.Date),
                pl.lit(ticker).alias("Ticker"),
                pl.lit(interval.value).alias("Interval"),
            ]
        )

        logger.info(f"Retrieved {len(df)} rows")
        return df

    def load_data(self, filepath: Union[str, Path]) -> pl.DataFrame:
        """
        Load existing data from file.

        Args:
            filepath: Path to the data file

        Returns:
            Polars DataFrame with raw data
        """
        filepath = Path(filepath)

        if not filepath.exists():
            raise FileNotFoundError(f"File not found: {filepath}")

        if filepath.suffix == ".csv":
            df = pl.read_csv(filepath)
        elif filepath.suffix == ".parquet":
            df = pl.read_parquet(filepath)
        else:
            raise ValueError(f"Unsupported format: {filepath.suffix}")

        # Ensure Date is proper type
        if "Date" in df.columns:
            df = df.with_columns(pl.col("Date").cast(pl.Date))

        return df

    def save_data(
        self,
        df: pl.DataFrame,
        filepath: Union[str, Path],
        format: FileFormat = FileFormat.CSV,
    ) -> None:
        """
        Save DataFrame to file.

        Args:
            df: Polars DataFrame to save
            filepath: Path where to save the file
            format: File format from FileFormat enum
        """
        if df.is_empty():
            raise ValueError("Cannot save empty DataFrame")

        filepath = Path(filepath)

        # Ensure extension matches format
        expected_ext = f".{format.value}"
        if filepath.suffix != expected_ext:
            filepath = filepath.with_suffix(expected_ext)

        # Create directory only when actually saving
        filepath.parent.mkdir(parents=True, exist_ok=True)

        if format == FileFormat.CSV:
            df.write_csv(filepath)
        elif format == FileFormat.PARQUET:
            df.write_parquet(filepath)
        else:
            raise ValueError(f"Unsupported format: {format}")

        logger.info(f"Saved to {filepath}")

    def fetch_data(
        self,
        ticker: str = "^GSPC",
        start_date: Optional[DateLike] = None,
        end_date: Optional[DateLike] = None,
        interval: TimeInterval = TimeInterval.DAILY,
        save: bool = False,
        filepath: Optional[Union[str, Path]] = None,
        format: FileFormat = FileFormat.CSV,
    ) -> pl.DataFrame:
        """
        Complete pipeline to fetch raw financial data - checks for existing file first.

        Args:
            ticker: Ticker symbol (e.g., 'SPY', '^GSPC')
            start_date: Start date as string 'YYYY-MM-DD' or datetime object
            end_date: End date as string 'YYYY-MM-DD' or datetime object
            interval: Data interval from TimeInterval enum
            save: Whether to save data to file
            filepath: Path where to save/load the file (auto-generated if None)
            format: File format from FileFormat enum

        Returns:
            Raw Polars DataFrame with OHLCV data
        """

        # Generate filepath if saving and none provided
        if save and filepath is None:
            start_str = self._parse_date(start_date, 365).replace("-", "")
            end_str = self._parse_date(end_date, 0).replace("-", "")
            filepath = f"data/raw/{ticker}_{start_str}_{end_str}_{interval.value}.{format.value}"

        # Try loading existing file first
        if filepath and Path(filepath).exists():
            try:
                df = self.load_data(filepath)
                logger.info(f"Loaded existing data from {filepath}")
                return df
            except Exception as e:
                logger.warning(f"Failed to load {filepath}: {e}")

        # Fetch fresh data
        df = self.get_data(ticker, start_date, end_date, interval)

        # Save if requested
        if save and filepath:
            self.save_data(df, filepath, format)

        return df


# Simple test/demo when running this file directly
if __name__ == "__main__":
    print("Testing DataFetcher:")
    fetcher = DataFetcher()
    df = fetcher.fetch_data(ticker="^GSPC", interval=TimeInterval.MONTHLY, save=False)
    print(f"Fetched {len(df)} rows for S&P 500. Latest 3:")
    print(df.tail(3))
